theano/compile/builders.py

```diff
@@ -271,21 +315,40 @@ def __str__(self):
         is_inline = self.is_inline
         return '%(name)s{inline=%(is_inline)s}' % locals()
 
-    def _recompute_grad_op(self):
+    def _recompute_lop_op(self):
         '''
-        converts self._grad_op from user supplied form to type(self) instance
+        converts self._lop_op from user supplied form to type(self) instance
 
         '''
         local_inputs = self.local_inputs
         local_outputs = self.local_outputs
         inp_len = len(local_inputs)
-        grad_op = self._grad_op
-
-        if isinstance(grad_op, OpFromGraph):
-            if not self._grad_op_is_cached:
-                self._grad_op_is_cached = True
-                self._grad_op_stypes_l = [None] * inp_len
-            return
+        lop_op = self._lop_op
+
+        if isinstance(lop_op, OpFromGraph):
+            if self._lop_op_is_cached:
+                return
+            if self._lop_type == 'grad':
+                needed_ninps = inp_len + len(local_outputs)
+                ninps = len(lop_op.local_inputs)
+                if needed_ninps != ninps:
+                    raise ValueError(
+                        self.OV_INP_LEN_ERR_MSG % (needed_ninps, ninps))
+                # make a wrapper callable
+                lop_op = lambda inps, grads: self._lop_op(*(inps + grads))  # flake8: noqa
+            elif self._lop_type == 'lop':
+                # OfG can be directly used in L_op format
+                needed_ninps = inp_len + 2 * len(local_outputs)
+                ninps = len(lop_op.local_inputs)
+                if needed_ninps != ninps:
+                    raise ValueError(
+                        self.OV_INP_LEN_ERR_MSG % (needed_ninps, ninps))
+                self._lop_op_is_cached = True
+                self._lop_op_stypes_l = [None] * inp_len
+                self._lop_op.kwargs['on_unused_input'] = 'ignore'
+                return
+            else:
+                raise ValueError(self.LOP_TYPE_ERR_MSG % self._lop_type)
```

Comment:
This should be an assert, not an exception since `_lop_type` is not user-settable and therefore should never be wrong.